### PR TITLE
Rust toolchain upgrade from 1.65 to 1.66

### LINKED ISSUE #3215: False negatives of `clippy::unneccesary_cast`/`clippy::needless_borrow`
Due to a bug in clippy that is fixed in nightly, the version of clippy we currently use does not warn for many instances of `clippy::unncessary_cast`: https://github.com/rust-lang/rust-clippy/issues/9562. Running `cargo +nightly clippy` on the firecracker workspace yields around 50 warnings of this type.

Something similar happens to a lesser extend with the `unnecessary_borrow` lint, however I could not track the change of behavior in clippy down to a specific issue.

This might cause issues if we ever decide to upgrade rust past 1.66 (once it is stable ofc), since we deny warnings.

### PR DESCRIPTION
**Pull request by andreitraistaru**
## Reason

Update Rust toolchain from 1.65 to 1.66. Fixing clippy errors (most of them related to `clippy::unecessary_cast`).

This PR should also fix #3215

## License Acceptance

By submitting this pull request, I confirm that my contribution is made under
the terms of the Apache 2.0 license.

## PR Checklist

- [x] All commits in this PR are signed (`git commit -s`).
- [x] If a specific issue led to this PR, this PR closes the issue.
- [x] The description of changes is clear and encompassing.
- [x] Any required documentation changes (code and docs) are included in this PR.
- [x] New `unsafe` code is documented.
- [x] API changes follow the [Runbook for Firecracker API changes][2].
- [ ] User-facing changes are mentioned in `CHANGELOG.md`.
- [x] All added/changed functionality is tested.
- [x] New `TODO`s link to an issue.

---

- [ ] This functionality can be added in [`rust-vmm`][1].

[1]: https://github.com/rust-vmm
[2]: https://github.com/firecracker-microvm/firecracker/blob/main/docs/api-change-runbook.md

### CODE REVIEW

**Review by bchalios — CHANGES_REQUESTED** [review]
Could you please split this in two commits? First commit should introduce the changes for upgrading the Rust version (Dockerfiles and devtool) and another one that fixes clippy errors.

**Comment by andreitraistaru** [comment]
Yes, sure, I'll do that right away. Also, I have to mention about the upgrade in the CHNAGELOG.

File: CHANGELOG.md [hunk 1]
```diff
@@ -5,6 +5,7 @@
 ### Changed
 
 - Improved error message when invalid network backend provided
+- Upgraded Rust toolchain from 1.64.0 to 1.66.0.
```
Comment (bchalios) [inline CHANGELOG.md:8]: Just double-checking that 1.64.0 is correct here. What was the latest toolchain that we released with?

**Review by JonathanWoollett-Light — APPROVED** [review]

File: CHANGELOG.md (hunk 1 above)
Comment (andreitraistaru) [reply CHANGELOG.md:8]: I upgraded from 1.65 to 1.66 but the PR for the upgrade from 1.64 to 1.65 didn't contain any CHANGELOG modifications so I wanted to fix that as well. I didn't want to have 2 separate lines for 1.64 -> 1.65 and 1.65 -> 1.66 as long as looked redundant to me. Should I change this into 2 separate changes, however ?

Comment (JonathanWoollett-Light) [reply CHANGELOG.md:8]: From the release commit for 1.2.0 we can see its 1.64.0.
https://github.com/firecracker-microvm/firecracker/blob/408f2fca1a3a55027245513531bd6abca8e39017/tools/devctr/Dockerfile.x86_64#L7

**Review by bchalios — APPROVED** [review]

File: CHANGELOG.md (hunk 1 above)
Comment (andreitraistaru) [reply CHANGELOG.md:8]: Yes, and on 5.12.2022 (which was after the 1.2 release that took place on 30.11.2022: https://github.com/firecracker-microvm/firecracker/commit/408f2fca1a3a55027245513531bd6abca8e39017), a PR for the upgrade to 1.65 has been merged: https://github.com/firecracker-microvm/firecracker/commit/4ae37a4d0896d73a8a684550e22f4440c10b94a4

Comment (andreitraistaru) [reply CHANGELOG.md:8]: That means the upgrade to 1.65 will be part of FC 1.3 and the changelog didn't contain anything related to rust upgrade 1.64->1.65 in the `Unreleased` section.

**Pull request merged**